### ISSUE TITLE
[BUG] <project> is missing dependency when using --scale service=0

### ISSUE THREAD

**Issue by envy** [docker/compose#11328]
### Description

We use compose to create a test environment in CI consisting of two database containers, two backends and a frontend.
See the following minimized and anonymized compose file:

```yml
version: '3.9'

services:
  database:
    image: postgres:15.2-alpine
    environment:
      POSTGRES_USER: ${POSTGRES_USER:-test}
      POSTGRES_HOST_AUTH_METHOD: trust
    ports:
      - '5432'
    healthcheck:
      test: ['CMD', 'pg_isready', '-U', '${POSTGRES_USER:-test}']
      interval: 10s
      timeout: 10s
      retries: 3
      start_period: 30s

  auditlog-database:
    image: timescale/timescaledb:latest-pg14
    environment:
      POSTGRES_USER: ${POSTGRES_USER:-test}
      POSTGRES_HOST_AUTH_METHOD: trust
    ports:
      - '5432'
    healthcheck:
      test: ['CMD', 'pg_isready', '-U', '${POSTGRES_USER:-test}']
      interval: 10s
      timeout: 10s
      retries: 3
      start_period: 30s

  backend:
    image: alpine
    depends_on:
      database:
        condition: service_healthy
      auditlog-database:
        condition: service_healthy

  auditlog:
    image: alpine
    depends_on:
      auditlog-database:
        condition: service_healthy

  frontend:
    image: alpine
    depends_on:
      - backend
      - auditlog
```

We updated compose from v2.22.0 to v2.23.3 and now we get a "<project> is missing dependency auditlog" error when starting with the following: `docker compose up -d --wait --scale backend=0 --scale auditlog=0 --scale frontend=0`

We do it this way as we then run `docker compose run backend ...` later with specific arguments to test specific parts of the system and reuse the same compose file in different test steps with different service scalings. Only this specific scaling does not work where all three are scaled down. Maybe others also don't work, we don't use all permutations possible.

### Steps To Reproduce

1. Copy the compose file above to `/somewhere`
2. run `docker compose up -d --wait --scale backend=0 --scale auditlog=0 --scale frontend=0`
3. See error

### Compose Version

```Text
/ # docker compose version
Docker Compose version v2.23.3
/ # docker-compose version
Docker Compose version v2.23.3
```

### Docker Environment

```Text
Client:
 Version:    24.0.7
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.0
    Path:     /usr/local/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.23.3
    Path:     /usr/local/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 10
  Running: 8
  Paused: 0
  Stopped: 2
 Images: 57
 Server Version: 24.0.6
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 61f9fd88f79f081d64d6fa3bb1a0dc71ec870523
 runc version: v1.1.9-0-gccaecfc
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.1.0-13-amd64
 Operating System: Debian GNU/Linux 12 (bookworm)
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 15.62GiB
 Name: devel
 ID: 903fd6e9-c5a8-4646-b2af-903f2ee214cf
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: true
 Insecure Registries:
  <redacted>:5000
  127.0.0.0/8
 Live Restore Enabled: false
```

### Anything else?

I'm guessing that the dependency resolution is not taking into account that the `frontend` service is scaled to 0 and therefore should not depend on anything (as nothing is actually started).

**Comment by ndeloof** [comment]
I was not able to reproduce this error running latest codebase. Would you have a chance to test latest release ?

**Comment by envy** [comment]
According to this repo, compose v2.23.3 should be the latest version which I did use for testing

Maybe a bit more info helps:

We use gitlab CI with a gitlab-runner that uses the `docker` image as the runner image, currently pinned to version 23.0.6 (which ships with compose v2.22.0).
The docker socket of the host is mounted inside the container started by the runner. 
The runner then issues the mentioned `docker compose up ...` command which did work fine.

The docker server on the host was updated to 24.0.7 and I updated the `docker` image tag to 24.0.7 as well so that the cli tools in the container match the host server version. Then, the runner was not able to issue the `docker compose up ...` command.

To test this locally on my machine I did this:

- Create a `compose-test` folder with the docker-compose.yml shown above
- `cd compose-test`
- `docker run --rm --entrypoint /bin/sh -v /var/run/docker.sock:/var/run/docker.sock -v .:/app -it docker:23.0.6`
- Inside the container run `cd /app` and `docker compose up -d --wait --scale backend=0 --scale auditlog=0 --scale frontend=0`. This will work
- Run `docker compose down` and exit the container
- Then run `docker run --rm --entrypoint /bin/sh -v /var/run/docker.sock:/var/run/docker.sock -v .:/app -it docker:24.0.7` (notice the different docker image tag)
- Again, `cd /app` and `docker compose up -d --wait --scale backend=0 --scale auditlog=0 --scale frontend=0`. This will now fail.
- Run `docker compose down` to clean up the database containers


I can try to build from the main branch and try that version of compose if you want.
